src/libopensc/card-edo2021.c

```diff
@@ -0,0 +1,275 @@
+		LOG_FUNC_RETURN(card->ctx, SC_ERROR_UNKNOWN);
+	}
+
+	LOG_FUNC_RETURN(card->ctx, SC_SUCCESS);
+}
+
+static int
+edo_set_security_env(struct sc_card *card, const struct sc_security_env *env, size_t data_len)
+{
+	SC_FUNC_CALLED(card->ctx, SC_LOG_DEBUG_VERBOSE);
+
+	if (!env)
+		LOG_FUNC_RETURN(card->ctx, SC_ERROR_INTERNAL);
+
+	if (env->algorithm != SC_ALGORITHM_EC || !(env->flags & SC_SEC_ENV_KEY_REF_PRESENT))
+		LOG_FUNC_RETURN(card->ctx, SC_ERROR_NOT_SUPPORTED);
+
+	u8 payload[] = {0x91, 0x02, 0x00, 0x00};
+
+	if (data_len <= SHA_DIGEST_LENGTH) {
+		payload[2] = 0x11;
+	} else if (data_len <= SHA256_DIGEST_LENGTH) {
+		payload[2] = 0x21;
+	} else if (data_len <= SHA512_DIGEST_LENGTH) {
+		payload[2] = 0x22;
+	} else {
+		sc_log(card->ctx, "unsupported data length for signature (%ld)", data_len);
+		LOG_FUNC_RETURN(card->ctx, SC_ERROR_NOT_SUPPORTED);
+	}
+
+	payload[3] |= env->key_ref[0];
+
+	struct sc_apdu apdu;
+	sc_format_apdu_ex(&apdu, 0x00, 0x22, 0x81, 0xB6, payload, sizeof payload, NULL, 0);
+
+	LOG_TEST_RET(card->ctx, sc_transmit_apdu(card, &apdu), "APDU transmit failed");
+	LOG_TEST_RET(card->ctx, sc_check_sw(card, apdu.sw1, apdu.sw2), "SW check failed");
+
+	LOG_FUNC_RETURN(card->ctx, SC_SUCCESS);
+}
+
+/*
+ * This function is a stub for driver calls, it caches data from for later execution of edo_set_security_env
+ * when signing.
+ */
+static int
+edo_set_security_env_driver_stub(struct sc_card *card, const struct sc_security_env *env, int se_num)
+{
+	SC_FUNC_CALLED(card->ctx, SC_LOG_DEBUG_VERBOSE);
+	struct edo2021_privdata *privdata = (struct edo2021_privdata *)card->drv_data;
+
+	if (!privdata)
+		LOG_FUNC_RETURN(card->ctx, SC_ERROR_INTERNAL);
+
+	if (env->operation != SC_SEC_OPERATION_SIGN || se_num > 0)
+		LOG_FUNC_RETURN(card->ctx, SC_ERROR_NOT_SUPPORTED);
+
+	memcpy(&privdata->sec_env, env, sizeof(struct sc_security_env));
+
+	LOG_FUNC_RETURN(card->ctx, SC_SUCCESS);
+}
+
+static int
+edo_compute_signature(struct sc_card *card, const u8 *data, size_t data_len, u8 *out, size_t outlen)
+{
+	SC_FUNC_CALLED(card->ctx, SC_LOG_DEBUG_VERBOSE);
+	struct edo2021_privdata *privdata = (struct edo2021_privdata *)card->drv_data;
+
+	if (!privdata)
+		LOG_FUNC_RETURN(card->ctx, SC_ERROR_INTERNAL);
+
+	/* Now we know the data length (from hashed input), perform the true set_security_env. */
+	LOG_TEST_RET(card->ctx, edo_set_security_env(card, &privdata->sec_env, data_len), "set_security_env failed");
+
+	/* Call default iso7816 handler, we need this card op only for security_env. */
+	LOG_FUNC_RETURN(card->ctx, sc_get_iso7816_driver()->ops->compute_signature(card, data, data_len, out, outlen));
+}
+
+static int
+edo_logout(sc_card_t *card)
+{
+	sc_sm_stop(card);
+	return SC_SUCCESS;
+}
+
+static int
+edo_init(sc_card_t *card)
+{
+	SC_FUNC_CALLED(card->ctx, SC_LOG_DEBUG_VERBOSE);
+
+	struct edo2021_privdata *privdata = (struct edo2021_privdata *)card->drv_data;
+	privdata = sc_mem_secure_alloc(sizeof(struct edo2021_privdata));
```

Comment:
I don't think this actually contains sensitive data so I suggest you simply use `calloc()`.